In [1]:
%%time
import pandas as pd

from sklearn.metrics import roc_auc_score, make_scorer, accuracy_score
from sklearn.preprocessing import OneHotEncoder, LabelEncoder
from sklearn.preprocessing import StandardScaler, Normalizer, MinMaxScaler


from sklearn.model_selection import train_test_split
from sklearn.model_selection import cross_val_score

from xgboost import XGBClassifier
from sklearn.feature_selection import RFE


from time import strftime, strptime, mktime, time, ctime

from sklearn import model_selection

from sklearn.linear_model import LogisticRegression
from sklearn.neighbors import KNeighborsClassifier
from sklearn.naive_bayes import GaussianNB 
from sklearn.svm import LinearSVC

from xgboost import XGBClassifier
import xgboost as xgb

from sklearn.tree import DecisionTreeClassifier, ExtraTreeClassifier
from sklearn.ensemble import AdaBoostClassifier, BaggingClassifier,\
ExtraTreesClassifier, GradientBoostingClassifier, VotingClassifier, RandomForestClassifier

from mlxtend.classifier import StackingClassifier, StackingCVClassifier
from mlxtend.feature_selection import ColumnSelector

from sklearn.pipeline import make_pipeline
import numpy as np

import gc


CPU times: user 568 ms, sys: 2.18 s, total: 2.75 s
Wall time: 534 ms


/home/shj16110/anaconda2/lib/python2.7/site-packages/sklearn/cross_validation.py:41: DeprecationWarning: This module was deprecated in version 0.18 in favor of the model_selection module into which all the refactored classes and functions are moved. Also note that the interface of the new CV iterators are different from that of this module. This module will be removed in 0.20.
  "This module will be removed in 0.20.", DeprecationWarning)


In [2]:
%%time
df = pd.read_csv('./train.csv')
print df.shape
test = pd.read_csv('./test.csv')
print test.shape

y = df.pop('Approved')
train_id = df.pop('ID')
test_id = test.pop('ID')

def helper(x):
    try:
        tmp = strptime(x, '%d/%m/%y')
        tmp = mktime(tmp)
    except:
        tmp = 563000400
    return tmp

df['Lead_Creation_Date'] = df['Lead_Creation_Date'].apply(helper)
test['Lead_Creation_Date'] = test['Lead_Creation_Date'].apply(helper)

df['DOB'] = df['DOB'].apply(helper)
test['DOB'] = test['DOB'].apply(helper)

df = df.fillna(axis=1, value=-1)
test = test.fillna(axis=1, value=-1)

df = pd.get_dummies(df)
test = pd.get_dummies(test)
test = test.reindex(columns = df.columns, fill_value=0)

test_o = test.copy()

(69713, 22)
(30037, 21)
CPU times: user 13.4 s, sys: 6.99 s, total: 20.4 s
Wall time: 20.4 s


In [3]:
%%time

rid = []

for i in df:
    if df[i].value_counts().head(1).values[0]>df.shape[0]*0.996:
        rid.append(i)
print len(rid), df.shape[1]-len(rid)

37317 99
CPU times: user 35.8 s, sys: 36 ms, total: 35.8 s
Wall time: 35.8 s


In [4]:
%%time
test = test_o.copy()
train = df.drop(rid, axis=1)
test = test.drop(rid, axis=1)

CPU times: user 1.71 s, sys: 2.26 s, total: 3.97 s
Wall time: 3.96 s


In [5]:
%%time
s = StandardScaler()
train = s.fit_transform(train)
test = s.transform(test)

print train.shape, test.shape

(69713, 99) (30037, 99)
CPU times: user 117 ms, sys: 80 ms, total: 197 ms
Wall time: 195 ms


### from here

In [26]:
train_test_split?

In [6]:
%%time
m = lambda *x: [i.shape for i in x]
X_train, X_test, y_train, y_test = train_test_split(train, y, test_size=0.2)
print m(X_train, y_train, X_test, y_test)

[(55770, 99), (55770,), (13943, 99), (13943,)]
CPU times: user 46 ms, sys: 33 ms, total: 79 ms
Wall time: 78.2 ms


In [7]:
%%time
d_train = xgb.DMatrix(X_train, label=y_train)
d_test = xgb.DMatrix(X_test, label=y_test)

CPU times: user 80 ms, sys: 18 ms, total: 98 ms
Wall time: 96.7 ms


In [8]:
RANDOM_STATE = 2018
params = {
    'min_child_weight': 0.8,
#     'eta': 0.1,
    'colsample_bytree': 0.8,
    'max_depth': 12,
    'subsample': 0.8,
    'alpha': 0.3,
    'lambda':0.9,
    'gamma': 0.3,
    'silent': 0,
    'learning_rate':0.02,
    'verbose_eval': True,
    'seed': RANDOM_STATE,
    'objective':"binary:logistic",
#     'eval_metric':'error'
#     'num_class':1
    'n_jobs':24,
    'eval_metric':'auc'
    
}
watchlist = [(d_train, 'train'), (d_test, 'test')]



In [18]:
start=time()

clf = xgb.train(params, d_train, 10000, watchlist, early_stopping_rounds=200, 
                verbose_eval=50,  maximize=True,
               )
                


'!!! used %.2f s'%(time()-start)

[0]	train-auc:0.74398	test-auc:0.771811
Multiple eval metrics have been passed: 'test-auc' will be used for early stopping.

Will train until test-auc hasn't improved in 200 rounds.
[50]	train-auc:0.8617	test-auc:0.851872
[100]	train-auc:0.912919	test-auc:0.863342
[150]	train-auc:0.95346	test-auc:0.871982
[200]	train-auc:0.97413	test-auc:0.872951
[250]	train-auc:0.984987	test-auc:0.874292
[300]	train-auc:0.991188	test-auc:0.874003
[350]	train-auc:0.994649	test-auc:0.872367
[400]	train-auc:0.99679	test-auc:0.870057
Stopping. Best iteration:
[234]	train-auc:0.982097	test-auc:0.874999



'!!! used 16.05 s'

In [25]:
# print set(df.columns)-set(rid)

[0]	train-auc:0.74398	test-auc:0.771811
Multiple eval metrics have been passed: 'test-auc' will be used for early stopping.

Will train until test-auc hasn't improved in 200 rounds.
[50]	train-auc:0.8617	test-auc:0.851872
[100]	train-auc:0.912919	test-auc:0.863342
[150]	train-auc:0.95346	test-auc:0.871982
[200]	train-auc:0.97413	test-auc:0.872951
[250]	train-auc:0.984987	test-auc:0.874292
[300]	train-auc:0.991188	test-auc:0.874003
[350]	train-auc:0.994649	test-auc:0.872367
[400]	train-auc:0.99679	test-auc:0.870057
Stopping. Best iteration:
[234]	train-auc:0.982097	test-auc:0.874999

Out[9]:
'!!! used 19.31 s'

In [17]:
df.columns

Index([u'DOB', u'Lead_Creation_Date', u'Employer_Category2', u'Monthly_Income',
       u'Existing_EMI', u'Loan_Amount', u'Loan_Period', u'Interest_Rate',
       u'EMI', u'Var1',
       ...
       u'Source_S160', u'Source_S161', u'Source_S162', u'Source_Category_A',
       u'Source_Category_B', u'Source_Category_C', u'Source_Category_D',
       u'Source_Category_E', u'Source_Category_F', u'Source_Category_G'],
      dtype='object', length=37416)

In [10]:
%%time
dtest = xgb.DMatrix(test)

CPU times: user 38 ms, sys: 5 ms, total: 43 ms
Wall time: 41.2 ms


In [11]:
%%time
predb = clf.predict(dtest)


CPU times: user 2.1 s, sys: 2 ms, total: 2.11 s
Wall time: 149 ms


___

In [12]:
submit = pd.concat([pd.DataFrame(test_id.values, columns=['ID']), pd.DataFrame(predb, columns=['Approved'])], axis=1)
submit.shape

(30037, 2)

In [14]:
submit.to_csv('./submit/xgb_rm996.csv',index=False)

#### lb .852388

In [60]:
# clf.save_model('xgb_0.852586.model')

In [441]:
roc_auc_score?

In [442]:
submit2 = pd.concat([pd.DataFrame(test_id.values, columns=['ID']), pd.DataFrame(pred, columns=['Approved'])], axis=1)
submit2.shape

(30037, 2)

In [449]:
accuracy_score(y_test, [0]*y_test.shape[0])

0.9834325467976762

___

In [276]:
cross_val_score(RandomForestClassifier(n_estimators=50, n_jobs=24), df.iloc[:,:-1], df.iloc[:,-1], cv=5)

array([0.87807502, 1.        , 0.99992828, 1.        , 0.98759145])

In [76]:
fr = RandomForestClassifier(n_estimators=50)
fr

RandomForestClassifier(bootstrap=True, class_weight=None, criterion='gini',
            max_depth=None, max_features='auto', max_leaf_nodes=None,
            min_impurity_decrease=0.0, min_impurity_split=None,
            min_samples_leaf=1, min_samples_split=2,
            min_weight_fraction_leaf=0.0, n_estimators=50, n_jobs=1,
            oob_score=False, random_state=None, verbose=0,
            warm_start=False)

In [77]:
fr.fit(df.iloc[:,:-1], df.iloc[:,-1])

RandomForestClassifier(bootstrap=True, class_weight=None, criterion='gini',
            max_depth=None, max_features='auto', max_leaf_nodes=None,
            min_impurity_decrease=0.0, min_impurity_split=None,
            min_samples_leaf=1, min_samples_split=2,
            min_weight_fraction_leaf=0.0, n_estimators=50, n_jobs=1,
            oob_score=False, random_state=None, verbose=0,
            warm_start=False)

In [ ]:
fr.predict_log_proba